src/script/interpreter.cpp

```diff
@@ -1213,6 +1342,185 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        // scripthash.
+                        if (stack.size() < 3) {
+                            return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                        }
+                        valtype& hash_from_stack = stacktop(-1);
+                        if (hash_from_stack.size() != 32) {
+                            return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                        }
+                        const uint256 target_hash{hash_from_stack};
+                        popstack(stack);
+
+                        if (!CheckUnvaultTriggerOutputsWitness(
+                                opuv_witversion, opuv_witprogram,
+                                recovery_spk_hash, spend_delay, target_hash,
+                                fRequireMinimal, flags, serror)) {
+                            return set_error(serror, SCRIPT_ERR_UNVAULT_MISMATCH);
+                        }
+                    }
+
+                    // Now that we have verified the structure of the outputs, verify
+                    // the signature authorizing the beginning of the unvault.
+                    //
+                    // Check that a valid trigger signature has been put on the witness
+                    // stack.
+                    const valtype& trigger_witness_program = stacktop(-1);
+                    auto trigger_witness_spk = CScript(
+                        trigger_witness_program.begin(), trigger_witness_program.end());
+
+                    const uint256 trigger_spk_hash = VaultScriptHash(
+                            HASHER_VAULT_UNVAULT_SPK, trigger_witness_spk);
+
+                    if (trigger_spk_hash != expected_trigger_spk_hash) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_WRONG_TRIGGER_WITNESS_PROGRAM);
+                    }
+                    popstack(stack);
+                    // Everything remaining on the stack is the witness stack
+                    // to be fed into the witness program above.
+
+                    CScriptWitness witness;
+                    witness.stack = stack;
+                    int witnessversion;
+                    std::vector<unsigned char> witnessprogram;
+
+                    if (!trigger_witness_spk.IsWitnessProgram(witnessversion, witnessprogram)) {
+                        return set_error(serror, SCRIPT_ERR_VAULT_INVALID_TRIGGER_WITNESS_PROGRAM);
+                    }
+                    // Note that this will recursively call EvalScript. We could get
+                    // into a further recursion if OP_VAULT spends are included in the
+                    // trigger witness program; the recursion depth is limited
+                    // solely by script size constraints.
+                    //
+                    // TODO: think more about whether to limit recursive OP_VAULT
+                    // evaluations in trigger witness programs.
```

Comment:
In fbf344931a2224dc3ace0011c1a4cb5cd4187f3e "wip: draft implementation"

I think it makes sense to disallow recursive OP_VAULTs in the OP_VAULT trigger script (? this was called unvault script in the paper). They necessarily have to have the same recovery spk hash and delay as the top level OP_VAULT, and at the end, would have some witness script which is what we actually care about at this point. So really all that would do is waste resources.

Reply:
Good point. This has been done; see the `limit_recursion` parameter. Thanks to @john-moffett for outlining a particular test case to ensure that recursion is limited to a single call.